### PR TITLE
Update of the minibook on building programs:

### PR DESCRIPTION
**Pull request by arjenmarkus**
- Text split into separate "chapters"/"sections"
- Applying consistent mark-up for code sections and keywords and such
- Applying most suggestions from the reviews (I have kept the "tabulate.f90"
  mostly as it was, but I have changed the name of the module to make it
  a bit clearer.
- Small textual changes in places

### CODE REVIEW

**Comment by milancurcic** [comment]
#build_preview

**Comment by milancurcic** [comment]
@arjenmarkus does this PR supersede the previous one (#90)?

**Comment by milancurcic** [comment]
@LKedward do you know why this preview didn't build? The preview site doesn't load for me.

**Comment by LKedward** [comment]
> @LKedward do you know why this preview didn't build? The preview site doesn't load for me.

It's working for me - can you try on another device, maybe with mobile internet?
I've had this issue before; I think there may be some kind of caching issue with github pages where if you visit the new address too soon, it caches the 404 and keeps serving you with it.

**Comment by LKedward** [comment]
@milancurcic - also give <https://fortran-lang.org/pr/99/index.html> a try.

**Comment by certik** [comment]
I think this looks great, thank you.

Once fpm is more usable, we can update the tutorial to show how to do this using fpm.

**Comment by arjenmarkus** [comment]
Hi Milan,

yes, this is the result of the suggestions made earlier by Jérémie and
others, and of splitting up the text based on the recipe by Laurence.

Regards,

Arjen

Op vr 5 jun. 2020 om 21:46 schreef Milan Curcic <notifications@github.com>:

File: learn/building_programs/linking_pieces.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+* it compiles the file "tabulate.f90" (using the module file);
+* it invokes the linker to combine the object files tabulate.o and function.o into an
+executable program - with the default name "a.out" or "a.exe" (if you
+want a different name, use the option "-o").
+
+What you do not see in general is that the linker also adds a number of
+extra files in this link step, the run-time libraries. These run-time
+libraries contain all the "standard" stuff - low-level routines that do
+the input and output to screen, the `sine` function and much more.
```
Comment (milancurcic) [inline learn/building_programs/linking_pieces.md:88]:
```suggestion
the input and output to screen, the `sin` function and much more.
```

File: learn/building_programs/managing_libraries.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+Building dynamic libraries works slightly differently from building
+static libraries: you use the compiler/linker instead of a tool
+like `ar` or `lib`.
+
+On Linux:
+
+```shell
+    $ gfortran -fpic -c file1.f90 file2.f90
+    $ gfortran -fpic -c file3.f90 ...
+    $ gfortran -shared --o supportlib.so file1.o file2.o file3.o ...
```
Comment (milancurcic) [inline learn/building_programs/managing_libraries.md:86]:
```suggestion
   $ gfortran -shared -o supportlib.so file1.o file2.o file3.o ...
```

File: learn/building_programs/managing_libraries.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+with checks on the function's interface, the right name and the reference
+to "a" DLL, called "function.dll".
+
+You can replace the shared library "function.dll" by another one, implementing
+a different function "f". Of course, you need to be careful to use the correct
+interface for this function. The compiler/linker are not invoked anymore, so they
+can do no checking.
+
+## Windows and Intel Fortran
+The setup is the same as with GNU/Linux, but on Windows it is necessary
```
Comment (milancurcic) [inline learn/building_programs/managing_libraries.md:181]:
```suggestion
The setup is the same as with Linux, but on Windows it is necessary
```

**Comment by milancurcic** [comment]
@arjenmarkus As you review the suggestions, click on "Commit suggestion" to automatically incorporate it in this PR, or just "Resolve conversation" to dismiss it.

File: learn/building_programs/include_files.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    tabulate/
+        main/
+            tabulate.f90
+        sub/
+            function.f90
+
+Compiling the file "function.f90" with the commands
+
+'''shell
```
Comment (everythingfunctional) [inline learn/building_programs/include_files.md:40]:
```suggestion
```shell

```

**Comment by arjenmarkus** [comment]
Hi everyone,

Thanks for the suggestions and the approval. I hope to find time in the
next few days (let's be careful with promises) to incorporate these
suggestions. (I confess being a trifle lazy when it comes to hunting down
the relevant URLs :)).

Regards,
Arjen


Op zo 7 jun. 2020 om 13:04 schreef Laurence Kedward <
notifications@github.com>:

**Comment by arjenmarkus** [comment]
Hi Milan,

wrt the indentation, hm, no, not really - at least I guess they are
superfluous now that the blocks are marked with the programming language. I
did not realise the indentation could be removed. That could certainly be a
cosmetic update.

Regards,

Arjen

Op za 6 jun. 2020 om 22:08 schreef Milan Curcic <notifications@github.com>:

File: learn/building_programs/distributing.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+---
+
+When you distributed your programs, there are a number of options you can
+choose from:
+
+1. Distribute the entire source code
+2. Distribute a pre-built executable program
+3. Distribute static or dynamic libraries that people can use
+
+ad 1.
```
[the remark this answers is not in the document]
Comment (arjenmarkus) [reply learn/building_programs/distributing.md:14]: Oh, I thought this kind of referencing was "universal" - "ad 1" means "look above for a concise description, here is a more extensive explanation". But that may be a neerlandicism ;).

**Comment by arjenmarkus** [comment]
How to resolve this conflict?

**Comment by arjenmarkus** [comment]
Hi everyone,

just did my best to get the remaining suggestions in. I do not know exactly
what to do with the conflict in learn.yml, so I would like your advice on
this.

Regards,

Arjen

Op ma 8 jun. 2020 om 08:45 schreef Arjen Markus <arjen.markus895@gmail.com>:

**Comment by LKedward** [comment]
Hi @arjenmarkus, many thanks.
I've resolved the conflict on github - you will need to run `git pull` on your local copy if you intend to do any more edits in your local copy.
I will update the #build_preview with the latest changes for a final check.

**Comment by arjenmarkus** [comment]
Hi @Lkedward, thanks - got the copy now. (I am still learning the workflow
:))

Regards,

Arjen

Op wo 17 jun. 2020 om 10:13 schreef Laurence Kedward <
notifications@github.com>:

File: learn/building_programs/index.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+---
+layout: book
+title: Building programs
+permalink: /learn/building_programs
+date: 2020-06-05
+author: Arjen Markus
```
Comment (milancurcic) [inline learn/building_programs/index.md:6]: Add co-authors to the list, sorted alphabetically by last name. Did I miss anybody? Please add your name if so.

```suggestion
author: Arjen Markus, Ondřej Čertík, Milan Curcic, Laurence Kedward, Brad Richardson, Jeremie Vandenplas
```

File: learn/building_programs/index.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+scripts.
+
+In this tutorial we concentrate on the first type of languages, with
+Fortran as the main example. One advantage of compiled languages is that
+the build process that you need to build an executable program, is used
+to transform the human-readable source code into an efficient program
+that can be run on the computer.
+
+Remark: this tutorial gives examples for the Windows and Linux operating
+systems, however the workflow and general principles still apply to MacOS.
```
Comment (milancurcic) [inline learn/building_programs/index.md:24]:
```suggestion
systems, however the workflow and general principles still apply to macOS.
```

**Comment by GS76** [comment]
Thank you for everything.

File: learn/building_programs/distributing.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        end module user_functions
+```
+
+* Provide a basic build script with a command like:
+
+```shell
+        gfortran -o function.dll function.f90 -shared
+```
+
+        or:
```
Comment (LKedward) [inline learn/building_programs/distributing.md:84]: For some reason this 'or' is being rendered in a code block, but I can't work out why?

File: learn/building_programs/distributing.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+---
+
+When you distribute your programs, there are a number of options you can
+choose from:
+
+1. Distribute the entire source code
+2. Distribute a pre-built executable program
+3. Distribute static or dynamic libraries that people can use
+
+ad 1.
```
Comment (LKedward) [inline learn/building_programs/distributing.md:14]:
```suggestion
__Option 1: Distribute the entire source code__
```

File: learn/building_programs/distributing.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+By far the simplest - for you as a programmer - is this one: you leave it
+up to the user to build it on their own machine. Unfortunately, that
+means you will have to have a user-friendly build system in place and
+the user will have to have access to suitable compilers. For build systems:
+see the previous section.
+
+ad 2.
```
Comment (LKedward) [inline learn/building_programs/distributing.md:22]:
```suggestion
__Option 2: Distribute a pre-built executable program__
```

File: learn/building_programs/distributing.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+Note: On Windows, the Intel Fortran comes with a set of _redistributable_ libraries.
+These will need to be made available.
+
+In general: use a tool like "ldd" or "dependency walker" to find out what
+libraries are required and consult the documentation of the compiler.
+
+If your program does allow customisation, consider using dynamic libraries for this.
+More is said about this below.
+
+ad 3.
```
Comment (LKedward) [inline learn/building_programs/distributing.md:45]:
```suggestion
__Option 3: Distribute static or dynamic libraries that people can use__
```

File: learn/building_programs/distributing.md (hunk 8 above)
Comment (LKedward) [reply learn/building_programs/distributing.md:84]: Okay, removing the indent fixes it.

```suggestion
or:
```

**Comment by arjenmarkus** [comment]
Hi everyone,

just committed the changes - I should now close the pull request, I guess?

Regards,

Arjen

Op wo 17 jun. 2020 om 15:46 schreef Laurence Kedward <
notifications@github.com>:

**Comment by LKedward** [comment]
Thanks @arjenmarkus - this looks good! Since you're happy with it, I will merge.

**Comment by LKedward** [comment]
#delete_preview

**Comment by arjenmarkus** [comment]
Nice to see this on the website :).

Regards,

Arjen

Op wo 17 jun. 2020 om 16:30 schreef github-actions[bot]
<notifications@github.com>:

**Pull request merged**